libnmstate/nm/wired.py

```diff
@@ -0,0 +1,147 @@
+#
+# Copyright 2018 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+import array
+import struct
+import fcntl
+import socket
+
+
+from libnmstate.nm import nmclient
+
+
+ETHTOOL_GSET = 0x00000001  # Get settings
+SIOCETHTOOL = 0x8946
+
+YES = 'yes'
+NO = 'no'
+
+
+def create_setting(iface_state, base_con_profile):
+    mtu = iface_state.get('mtu')
+
+    ethernet = iface_state.get('ethernet', {})
+    speed = ethernet.get('speed')
+    duplex = ethernet.get('duplex')
+    auto_negotiation = ethernet.get('auto-negotiation')
+
+    if not (mtu or speed or duplex or auto_negotiation):
+        return None
+
+    if auto_negotiation == YES:
+        auto_negotiation = True
+
+    if auto_negotiation == NO:
+        auto_negotiation = False
+
+    wired_setting = None
+    if base_con_profile:
+        wired_setting = base_con_profile.get_setting_wired()
+        if wired_setting:
+            wired_setting = wired_setting.duplicate()
+
+    if not wired_setting:
+        wired_setting = nmclient.NM.SettingWired.new()
+
+    if mtu:
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_MTU, mtu)
+
+    if speed:
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_SPEED, speed)
+
+    if duplex:
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_DUPLEX, duplex)
+
+    if auto_negotiation is not None:
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_AUTO_NEGOTIATE,
+                                   auto_negotiation)
+
+    return wired_setting
+
+
+# based on:
+# https://github.com/rlisagor/pynetlinux/blob/master/pynetlinux/ifconfig.py
+# https://elixir.bootlin.com/linux/v4.19-rc1/source/include/uapi/linux/ethtool.h
+def minimal_ethtool(interface):
```

Comment:
Could you create a dedicated module for this? ethtool.py under libnmstate would be nice.
Also, please review VDSM ethtool module, it includes socket protection (making sure it is closed):
https://github.com/oVirt/vdsm/blob/master/lib/vdsm/network/ethtool.py

Reply:
Do I Need `IFREQ_FORMAT = '16sPi'` - it works with `16sP` currently without the buffer length. I do not see any indication that there should be a buffer length. I added a finally statement to close the socket.